### ISSUE TITLE
`allow(clippy::duplicated_attributes)` does not take effect across out-of-line modules

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#12538]
### Summary

For this particular lint, module-level `#![allow(clippy::duplicated_attributes)]` suppresses occurrences of the lint within that module, and within in-line modules in the same file, but not in modules from other files.

This behavior is unlike what happens with `allow` for all other clippy lints, in which it applies to the whole module hierarchy across files.

### Reproducer

```rust
// src/lib.rs

#![allow(clippy::duplicated_attributes)]

pub mod here {
    #[allow(dead_code)]
    #[allow(dead_code)]
    pub fn foo() {}
}

pub mod there;
```

```rust
// src/there.rs

#[allow(dead_code)]
#[allow(dead_code)]
pub fn foo() {}
```

```console
$ cargo clippy
warning: duplicated attribute
 --> src/there.rs:4:9
  |
4 | #[allow(dead_code)]
  |         ^^^^^^^^^
  |
note: first defined here
 --> src/there.rs:3:9
  |
3 | #[allow(dead_code)]
  |         ^^^^^^^^^
help: remove this attribute
 --> src/there.rs:4:9
  |
4 | #[allow(dead_code)]
  |         ^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#duplicated_attributes
  = note: `#[warn(clippy::duplicated_attributes)]` on by default
```

The occurrence within `mod here` is suppressed by the `allow` but the one within `mod there` is not.

### Version

```text
rustc 1.79.0-nightly (85e449a32 2024-03-22)
binary: rustc
commit-hash: 85e449a3237e82c9ade8936a82bd4fc64cfe1057
commit-date: 2024-03-22
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2
```

**Comment by dtolnay** [comment]
@GuillaumeGomez @blyxyas (the lint is new from https://github.com/rust-lang/rust-clippy/pull/12378)

**Comment by J-ZhengLi** [comment]
It appears to be another victim of `early pass`, just like the one lint from #12436

**Comment by y21** [comment]
we should also probably add a comment here saying that you should prefer adding the lint to `Attributes` (instead of `EarlyAttributes`) if at all possible
https://github.com/rust-lang/rust-clippy/blob/4a8c9495ca72ffc4e25ec75984d5678c5b9ae796/clippy_lints/src/attrs/mod.rs#L588

**Comment by jdonszelmann** [comment]
fixed, this is no longer a pre-expansion lint after some recent changes of mine for the attribute rework (@y21)

**Comment by dtolnay** [comment]
Using `cargo bisect-rustc --component=clippy --regress=success -- clippy`, the fix bisects to nightly-2024-04-20. Within the PRs that day, probably https://github.com/rust-lang/rust/pull/124123, and within that, probably https://github.com/rust-lang/rust-clippy/pull/12646.
